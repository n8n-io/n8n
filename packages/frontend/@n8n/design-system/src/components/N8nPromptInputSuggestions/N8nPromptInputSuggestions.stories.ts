import type { StoryFn } from '@storybook/vue3-vite';

import N8nPromptInputSuggestions from './N8nPromptInputSuggestions.vue';
import type { WorkflowSuggestion } from '../../types/assistant';
import N8nPromptInput from '../N8nPromptInput';

export default {
	title: 'Assistant/PromptInputSuggestions',
	component: N8nPromptInputSuggestions,
	argTypes: {
		suggestionClick: { action: 'suggestion-clicked' },
	},
};

const mockSuggestions: WorkflowSuggestion[] = [
	{
		id: 'invoice-pipeline',
		summary: 'Invoice processing pipeline',
		prompt:
			'Create an invoice parsing workflow using n8n forms. Extract key information and store in Airtable.',
	},
	{
		id: 'ai-news-digest',
		summary: 'Daily AI news digest',
		prompt:
			'Create a workflow that fetches the latest AI news every morning at 8 AM and sends a summary via Telegram.',
	},
	{
		id: 'rag-assistant',
		summary: 'RAG knowledge assistant',
		prompt:
			'Build a pipeline that accepts PDF files, chunks documents, and creates a chatbot that can answer questions.',
	},
	{
		id: 'email-summary',
		summary: 'Summarize emails with AI',
		prompt:
			'Build a workflow that retrieves emails, performs AI analysis, and sends a summary to Slack.',
	},
];

const Template: StoryFn = (args) => ({
	setup: () => ({ args }),
	components: {
		N8nPromptInputSuggestions,
		N8nPromptInput,
	},
	template: `
		<div style="max-width: 710px; margin: 0 auto; padding: 20px;">
			<N8nPromptInputSuggestions v-bind="args">
				<template #prompt-input>
					<N8nPromptInput
						placeholder="Describe the workflow you want to build..."
						:min-lines="2"
						:streaming="args.streaming"
						:disabled="args.disabled"
						:credits-quota="args.creditsQuota"
						:credits-remaining="args.creditsRemaining"
						:show-ask-owner-tooltip="args.showAskOwnerTooltip"
						@submit="() => {}"
						@upgrade-click="() => {}"
					/>
				</template>
			</N8nPromptInputSuggestions>
		</div>
	`,
});

export const Default = Template.bind({});
Default.args = {
	suggestions: mockSuggestions,
};

export const Disabled = Template.bind({});
Disabled.args = {
	suggestions: mockSuggestions,
	disabled: true,
};

export const Streaming = Template.bind({});
Streaming.args = {
	suggestions: mockSuggestions,
	streaming: true,
};

export const NoSuggestions = Template.bind({});
NoSuggestions.args = {
	suggestions: [],
};

export const WithCredits = Template.bind({});
WithCredits.args = {
	suggestions: mockSuggestions,
	creditsQuota: 100,
	creditsRemaining: 75,
};
