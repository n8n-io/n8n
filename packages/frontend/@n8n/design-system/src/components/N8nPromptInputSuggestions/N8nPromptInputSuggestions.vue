<script setup lang="ts">
import type { WorkflowSuggestion } from '../../types/assistant';

interface Props {
	suggestions?: WorkflowSuggestion[];
	disabled?: boolean;
	streaming?: boolean;
}

const emit = defineEmits<{
	suggestionClick: [suggestion: WorkflowSuggestion];
}>();

const props = withDefaults(defineProps<Props>(), {
	suggestions: () => [],
	disabled: false,
	streaming: false,
});

function onSuggestionClick(suggestion: WorkflowSuggestion) {
	if (props.disabled || props.streaming) return;
	emit('suggestionClick', suggestion);
}
</script>

<template>
	<article :class="$style.container" data-test-id="workflow-suggestions">
		<!-- Prompt input slot -->
		<section :class="$style.promptContainer">
			<slot name="prompt-input" />
		</section>

		<!-- Suggestion pills section -->
		<section
			v-if="suggestions.length > 0 && !streaming"
			:class="$style.pillsContainer"
			role="group"
			aria-label="Workflow suggestions"
		>
			<button
				v-for="suggestion in suggestions"
				:key="suggestion.id"
				:class="$style.suggestionPill"
				:disabled="streaming || disabled"
				type="button"
				@click="onSuggestionClick(suggestion)"
			>
				{{ suggestion.summary }}
			</button>
		</section>

		<!-- Footer slot for optional content like "or" divider -->
		<slot name="footer" />
	</article>
</template>

<style lang="scss" module>
/* Layout */
.container {
	display: flex;
	flex-direction: column;
	align-items: center;
	gap: var(--spacing-m);
	max-width: 710px;
	width: 100%;
}

/* Prompt Input Section */
.promptContainer {
	display: flex;
	flex-direction: column;
	gap: var(--spacing-s);
	align-self: stretch;
}

/* Suggestion Pills Section */
.pillsContainer {
	display: flex;
	justify-content: center;
	align-items: flex-start;
	gap: var(--spacing-2xs);
	flex-wrap: wrap;
}

.suggestionPill {
	display: flex;
	padding: var(--spacing-4xs) var(--spacing-2xs);
	justify-content: center;
	align-items: center;
	gap: var(--spacing-2xs);
	border-radius: 56px;
	border: var(--border-base);
	background: var(--color-background-xlight);
	cursor: pointer;
	transition: all 0.2s ease;
	font-size: var(--font-size-2xs);
	color: var(--color-text-dark);
	font-weight: var(--font-weight-regular);

	&:hover:not(:disabled) {
		color: var(--color-primary);
		border-color: var(--color-primary);
		background: var(--color-background-xlight);
	}

	&:disabled {
		cursor: not-allowed;
		opacity: 0.6;
	}
}
</style>
