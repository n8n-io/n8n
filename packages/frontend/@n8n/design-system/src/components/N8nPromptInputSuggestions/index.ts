import N8nPromptInputSuggestions from './N8nPromptInputSuggestions.vue';

export default N8nPromptInputSuggestions;
export type { WorkflowSuggestion } from '../../types/assistant';
