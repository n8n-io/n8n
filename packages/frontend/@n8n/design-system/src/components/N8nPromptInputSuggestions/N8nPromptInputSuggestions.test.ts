import { render } from '@testing-library/vue';
import { describe, it, expect } from 'vitest';

import N8nPromptInputSuggestions from './N8nPromptInputSuggestions.vue';
import type { WorkflowSuggestion } from '../../types/assistant';

const mockSuggestions: WorkflowSuggestion[] = [
	{
		id: 'test-1',
		summary: 'Test Suggestion 1',
		prompt: 'This is test prompt 1',
	},
	{
		id: 'test-2',
		summary: 'Test Suggestion 2',
		prompt: 'This is test prompt 2',
	},
	{
		id: 'test-3',
		summary: 'Test Suggestion 3',
		prompt: 'This is test prompt 3',
	},
];

describe('N8nPromptInputSuggestions', () => {
	it('renders correctly with default props', () => {
		const { container } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: mockSuggestions,
			},
		});
		expect(container).toMatchSnapshot();
	});

	it('renders all suggestion pills', () => {
		const { getByText } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: mockSuggestions,
			},
		});
		mockSuggestions.forEach((suggestion) => {
			expect(getByText(suggestion.summary)).toBeTruthy();
		});
	});

	it('emits suggestion-click event when pill is clicked', () => {
		const { getByText, emitted } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: mockSuggestions,
			},
		});

		const firstPill = getByText(mockSuggestions[0].summary) as unknown as HTMLInputElement;
		firstPill.click();

		expect(emitted()).toHaveProperty('suggestionClick');
		expect(emitted().suggestionClick[0]).toEqual([mockSuggestions[0]]);
	});

	it('hides suggestion pills when streaming', () => {
		const { container } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: mockSuggestions,
				streaming: true,
			},
		});

		const pillsContainer = container.querySelector('[role="group"]');
		expect(pillsContainer).toBeFalsy();
	});

	it('disables pills when disabled prop is true', () => {
		const { getByText } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: mockSuggestions,
				disabled: true,
			},
		});

		const firstPill = getByText(mockSuggestions[0].summary) as unknown as HTMLInputElement;
		expect(firstPill.disabled).toBe(true);
	});

	it('does not render suggestion pills when suggestions array is empty', () => {
		const { container } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: [],
			},
		});

		const pillsContainer = container.querySelector('[role="group"]');
		expect(pillsContainer).toBeFalsy();
	});

	it('renders prompt-input slot content', () => {
		const { getByTestId } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: mockSuggestions,
			},
			slots: {
				'prompt-input': '<div data-test-id="custom-prompt">Custom Prompt Input</div>',
			},
		});

		expect(getByTestId('custom-prompt')).toBeTruthy();
	});

	it('renders footer slot content', () => {
		const { getByTestId } = render(N8nPromptInputSuggestions, {
			props: {
				suggestions: mockSuggestions,
			},
			slots: {
				footer: '<div data-test-id="custom-footer">Custom Footer</div>',
			},
		});

		expect(getByTestId('custom-footer')).toBeTruthy();
	});
});
