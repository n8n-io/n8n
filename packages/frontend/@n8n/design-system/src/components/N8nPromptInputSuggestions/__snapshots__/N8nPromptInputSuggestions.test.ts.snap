// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`N8nPromptInputSuggestions > renders correctly with default props 1`] = `
<div>
  <article
    class="container"
    data-test-id="workflow-suggestions"
  >
    <!-- Prompt input slot -->
    <section
      class="promptContainer"
    >
      
      
    </section>
    <!-- Suggestion pills section -->
    <section
      aria-label="Workflow suggestions"
      class="pillsContainer"
      role="group"
    >
      
      <button
        class="suggestionPill"
        type="button"
      >
        Test Suggestion 1
      </button>
      <button
        class="suggestionPill"
        type="button"
      >
        Test Suggestion 2
      </button>
      <button
        class="suggestionPill"
        type="button"
      >
        Test Suggestion 3
      </button>
      
    </section>
    <!-- Footer slot for optional content like "or" divider -->
    
    
  </article>
</div>
`;
